File: 2022/04/day4.go
package main

import (
	"fmt"
	"os"
	"strconv"
	"strings"

	"AoC/utils"
)

type Range struct {
	min int
	max int
}

func (r Range) includes(val int) bool {
	return val >= r.min && val <= r.max
}

func (r Range) contains(other Range) bool {
	return r.includes(other.min) && r.includes(other.max)
}

func (r Range) overlaps(other Range) bool {
	return r.includes(other.min) || r.includes(other.max)
}

func makeRange(rangeStr string) Range {
	ns := strings.Split(rangeStr, "-")
	min := utils.CheckErr(strconv.Atoi(ns[0]))
	max := utils.CheckErr(strconv.Atoi(ns[1]))
	return Range{min, max}
}

func main() {
	input := utils.CheckErr(os.ReadFile("input.txt"))

	var part1, part2 int
	for _, line := range utils.Lines(string(input)) {
		pair := strings.Split(line, ",")
		r1 := makeRange(pair[0])
		r2 := makeRange(pair[1])
		if r1.contains(r2) || r2.contains(r1) {
			part1 += 1
		}
		if r1.overlaps(r2) || r2.overlaps(r1) {
			part2 += 1
		}
	}

	fmt.Printf("The answer to Part 1 is %v.\n", part1)
	fmt.Printf("The answer to Part 2 is %v.\n", part2)
}

File: 2015/04/day4.py
from hashlib import md5


with open("input.txt") as f:
    key = f.readline().strip()

i = 1
part1 = None
part2 = None
while True:
    input = (key + str(i)).encode()
    hex_hash = md5(input).hexdigest()
    if not part1 and hex_hash.startswith("00000"):
        part1 = i
    if not part2 and hex_hash.startswith("000000"):
        part2 = i
    if part1 and part2:
        break
    i += 1

print(f"The answer to Part 1 is {part1}.")
print(f"The answer to Part 2 is {part2}.")
